futures-util/src/future/join_all.rs

```diff
@@ -86,27 +134,20 @@ where
     ) -> Poll<Self::Output> {
         let mut all_done = true;
 
-        for elem in self.as_mut().elems.iter_mut() {
-            match elem {
-                ElemState::Pending(ref mut t) => match Pin::new(t).poll(lw) {
-                    Poll::Ready(v) => *elem = ElemState::Done(v),
-                    Poll::Pending => {
-                        all_done = false;
-                        continue;
-                    }
-                },
-                ElemState::Done(ref mut _v) => (),
-            };
+        for mut elem in iter_pin_mut(self.elems.as_mut()) {
+            if let Some(pending) = elem.as_mut().pending_pin_mut() {
+                if let Poll::Ready(output) = pending.poll(lw) {
+                    elem.set(ElemState::Done(Some(output)));
+                } else {
+                    all_done = false;
+                }
+            }
         }
 
         if all_done {
-            let elems = mem::replace(&mut self.elems, Vec::new());
-            let result = elems
-                .into_iter()
-                .map(|e| match e {
-                    ElemState::Done(t) => t,
-                    _ => unreachable!(),
-                })
+            let mut elems = mem::replace(&mut self.elems, Box::pin([]));
```

Comment:
Isn't that causing a new allocation to keep `elems` intact, even though we don't need it anymore? If yes it would be nice to avoid it.

Reply:
Hmm, does `Box` allocate for a dynamic ZST? If so I guess `self.elems` could be an `Option`, but I'm hoping `Box` is smart enough to not actually allocate in this case (I'll see if I can come up with a testcase showing it).

Reply:
Oh it’s zero sized? I’m not sure. Actually it wouldn’t make sense, because malloc(0) is borderline invalid (not super well defined). 

Guess it might be checkable with global allocator?

Reply:
Oh it’s zero sized? I’m not sure. Actually it wouldn’t make sense, because malloc(0) is borderline invalid (not super well defined). 

Guess it might be checkable with global allocator?

Reply:
Heh, should have just checked [the docs](https://doc.rust-lang.org/nightly/std/boxed/struct.Box.html#method.new):

> This doesn't actually allocate if T is zero-sized.

So it should just be basically updating the capacity to zero.